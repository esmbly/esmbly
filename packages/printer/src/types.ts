export type Callback = (error: Error | null | undefined) => void;

export type Encoding = string | undefined;

export type WriteFn = (
  chunk: string,
  encoding?: Encoding,
  callback?: Callback,
) => boolean;
