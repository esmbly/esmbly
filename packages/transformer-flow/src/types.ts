import t from '@babel/types';
import { Node, Visitor } from '@babel/traverse';

export interface Warning {
  info: string;
  issueUrl: string;
  node: t.Node;
}

type RuleA = () => Visitor<Node>;
type RuleB = (warnings: Warning[]) => Visitor<Node>;
export type Rule = RuleA | RuleB;
