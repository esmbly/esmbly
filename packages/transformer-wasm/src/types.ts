export interface WasmTransformerOptions {
  optimize?: string;
  optimizeLevel?: number;
  shrinkLevel?: number;
  validate?: boolean;
  // TODO: ADD more options
}
