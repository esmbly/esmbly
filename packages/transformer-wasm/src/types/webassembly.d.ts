/*
WebAssembly v1 (MVP) declaration file for TypeScript
Definitions by: 01alchemist (https://twitter.com/01alchemist)

MIT License

Copyright (c) 2017 01Alchemist

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
*/

declare namespace WebAssembly {
  /**
   * WebAssembly.Module
   **/
  class Module {
      constructor (bufferSource: ArrayBuffer | Uint8Array);

      static customSections(module: Module, sectionName: string): ArrayBuffer[];
      static exports(module: Module): {
          name: string;
          kind: string;
      }[];
      static imports(module: Module): {
          module: string;
          name: string;
          kind: string;
      }[];
  }

  /**
   * WebAssembly.Instance
   **/
  class Instance {
      readonly exports: any;
      constructor (module: Module, importObject?: any);
  }

  /**
   * WebAssembly.Memory
   * Note: A WebAssembly page has a constant size of 65,536 bytes, i.e., 64KiB.
   **/
  interface MemoryDescriptor {
      initial: number;
      maximum?: number;
  }

  class Memory {
      readonly buffer: ArrayBuffer;
      constructor (memoryDescriptor: MemoryDescriptor);
      grow(numPages: number): number;
  }

  /**
   * WebAssembly.Table
   **/
  interface TableDescriptor {
      element: "anyfunc",
      initial: number;
      maximum?: number;
  }

  class Table {
      readonly length: number;
      constructor (tableDescriptor: TableDescriptor);
      get(index: number): Function;
      grow(numElements: number): number;
      set(index: number, value: Function): void;
  }

  /**
   * Errors
   */
  class CompileError extends Error {
      readonly fileName: string;
      readonly lineNumber: string;
      readonly columnNumber: string;
      constructor (message?:string, fileName?:string, lineNumber?:number);
      toString(): string;
  }

  class LinkError extends Error {
      readonly fileName: string;
      readonly lineNumber: string;
      readonly columnNumber: string;
      constructor (message?:string, fileName?:string, lineNumber?:number);
      toString(): string;
  }

  class RuntimeError extends Error {
      readonly fileName: string;
      readonly lineNumber: string;
      readonly columnNumber: string;
      constructor (message?:string, fileName?:string, lineNumber?:number);
      toString(): string;
  }

  function compile(bufferSource: ArrayBuffer | Uint8Array): Promise<Module>;

  interface ResultObject {
      module: Module;
      instance: Instance;
  }

  function instantiate(bufferSource: ArrayBuffer | Uint8Array, importObject?: any): Promise<ResultObject>;
  function instantiate(module: Module, importObject?: any): Promise<Instance>;

  function validate(bufferSource: ArrayBuffer | Uint8Array): boolean;
}