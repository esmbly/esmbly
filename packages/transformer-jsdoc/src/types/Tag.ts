export interface TagType {
  type: string;
  name: string;
  elements?: TagType[];
  expression: TagType;
  applications: TagType[];
}

export interface Tag {
  title: string;
  description: string;
  type: TagType;
  name: string;
}
