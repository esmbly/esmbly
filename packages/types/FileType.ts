export enum FileType {
  JavaScript = 'JavaScript',
  Flow = 'Flow',
  TypeScript = 'TypeScript',
  WebAssembly = 'WebAssembly',
}
