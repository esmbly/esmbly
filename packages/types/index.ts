export * from './AST';
export * from './File';
export * from './FileType';
export * from './RunConfig';
export * from './Transformer';
