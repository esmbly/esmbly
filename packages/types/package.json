{
  "name": "@esmbly/types",
  "version": "0.0.6",
  "description": "Type definitions for Esmbly, a composable transpiler infrastructure for working with statically typed JavaScript and WebAssembly",
  "author": "Olle Lauri Bostr√∂m <ollebostr@gmail.com> (https://github.com/ollelauribostrom)",
  "homepage": "https://github.com/esmbly/esmbly",
  "license": "MIT",
  "main": "dist/index.js",
  "types": "dist/index.d.ts",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/esmbly/esmbly.git"
  },
  "bugs": {
    "url": "https://github.com/esmbly/esmbly/issues"
  },
  "scripts": {
    "build": "tsc",
    "build-watch": "tsc -w"
  },
  "dependencies": {
    "@babel/parser": "^7.4.3",
    "@babel/traverse": "^7.4.3",
    "@babel/types": "^7.3.4"
  },
  "publishConfig": {
    "access": "public"
  }
}
