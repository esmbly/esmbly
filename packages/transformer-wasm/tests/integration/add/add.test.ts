import { Format } from '@esmbly/types';
import testRunner from '../../helpers/testRunner';

const program = `
  export function add(a: number, b: number): number {
    return a + b;
  }
`;

describe('integration: add', () => {
  it('transforms the input program to WebAssembly', async () => {
    const output = [{ format: Format.WebAssembly }];
    const [{ content }] = await testRunner(program, output);
    // @ts-ignore
    const { instance } = await WebAssembly.instantiate(content, {});
    expect(instance.exports.add(2, 3)).toEqual(5);
  });

  it('transforms the input program to Wat', async () => {
    const output = [{ format: Format.Wat }];
    const [{ content }] = await testRunner(program, output);
    expect(content).toMatchSnapshot();
  });

  it('transforms the input program to asm.js', async () => {
    const output = [{ format: Format.Asm }];
    const [{ content }] = await testRunner(program, output);
    expect(content).toMatchSnapshot();
  });

  it('transforms the input program to AssemblyScript', async () => {
    const output = [{ format: Format.AssemblyScript }];
    const [{ content }] = await testRunner(program, output);
    expect(content).toMatchSnapshot();
  });
});
