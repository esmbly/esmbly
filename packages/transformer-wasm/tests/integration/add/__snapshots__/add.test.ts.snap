// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`integration: add transforms the input program to AssemblyScript 1`] = `
"
  export function add(a: f64, b: f64): f64 {
    return a + b;
  }
"
`;

exports[`integration: add transforms the input program to Wat 1`] = `
"(module
 (type $FUNCSIG$ddd (func (param f64 f64) (result f64)))
 (type $FUNCSIG$v (func))
 (memory $0 0)
 (table $0 1 funcref)
 (elem (i32.const 0) $null)
 (global $~lib/memory/HEAP_BASE i32 (i32.const 8))
 (export \\"memory\\" (memory $0))
 (export \\"table\\" (table $0))
 (export \\"add\\" (func $testfile/add))
 (func $testfile/add (; 0 ;) (type $FUNCSIG$ddd) (param $0 f64) (param $1 f64) (result f64)
  local.get $0
  local.get $1
  f64.add
 )
 (func $null (; 1 ;) (type $FUNCSIG$v)
 )
)
"
`;

exports[`integration: add transforms the input program to asm.js 1`] = `
"function asmFunc(global, env, buffer) {
 \\"almost asm\\";
 var HEAP8 = new global.Int8Array(buffer);
 var HEAP16 = new global.Int16Array(buffer);
 var HEAP32 = new global.Int32Array(buffer);
 var HEAPU8 = new global.Uint8Array(buffer);
 var HEAPU16 = new global.Uint16Array(buffer);
 var HEAPU32 = new global.Uint32Array(buffer);
 var HEAPF32 = new global.Float32Array(buffer);
 var HEAPF64 = new global.Float64Array(buffer);
 var Math_imul = global.Math.imul;
 var Math_fround = global.Math.fround;
 var Math_abs = global.Math.abs;
 var Math_clz32 = global.Math.clz32;
 var Math_min = global.Math.min;
 var Math_max = global.Math.max;
 var Math_floor = global.Math.floor;
 var Math_ceil = global.Math.ceil;
 var Math_sqrt = global.Math.sqrt;
 var abort = env.abort;
 var nan = global.NaN;
 var infinity = global.Infinity;
 var $lib_memory_HEAP_BASE = 8;
 var i64toi32_i32$HIGH_BITS = 0;
 function testfile_add($0, $1) {
  $0 = +$0;
  $1 = +$1;
  return +($0 + $1);
 }
 
 function null_() {
  
 }
 
 var FUNCTION_TABLE_v = [null_];
 function __wasm_grow_memory(pagesToAdd) {
  pagesToAdd = pagesToAdd | 0;
  var oldPages = __wasm_current_memory() | 0;
  var newPages = oldPages + pagesToAdd | 0;
  if ((oldPages < newPages) && (newPages < 65536)) {
   var newBuffer = new ArrayBuffer(Math_imul(newPages, 65536));
   var newHEAP8 = new global.Int8Array(newBuffer);
   newHEAP8.set(HEAP8);
   HEAP8 = newHEAP8;
   HEAP16 = new global.Int16Array(newBuffer);
   HEAP32 = new global.Int32Array(newBuffer);
   HEAPU8 = new global.Uint8Array(newBuffer);
   HEAPU16 = new global.Uint16Array(newBuffer);
   HEAPU32 = new global.Uint32Array(newBuffer);
   HEAPF32 = new global.Float32Array(newBuffer);
   HEAPF64 = new global.Float64Array(newBuffer);
   buffer = newBuffer;
  }
  return oldPages;
 }
 
 function __wasm_current_memory() {
  return buffer.byteLength / 65536 | 0;
 }
 
 return {
  memory: Object.create(Object.prototype, {
   grow: {
    value: __wasm_grow_memory
   }, 
   buffer: {
    get: function () {
     return buffer;
    }
    
   }
  }), 
  add: testfile_add
 };
}

const memasmFunc = new ArrayBuffer(65536);
const retasmFunc = asmFunc({Math,Int8Array,Uint8Array,Int16Array,Uint16Array,Int32Array,Uint32Array,Float32Array,Float64Array,NaN,Infinity}, {abort:function() { throw new Error('abort'); }},memasmFunc);
export const memory = retasmFunc.memory;
"
`;
